docs/designs/keras_callback.md

```diff
@@ -0,0 +1,208 @@
+
+This document describes the design for supporting callback to customize the behavior of model during training, evaluation and inference in ElasticDL.
+
+## Motivation
+
+In deep learning, we generally need to customize the behavior of model during training, evaluation or
+inference, including reading/changing the model. According to the call frequency, we may perform the behavior per batch, per epoch or at the start and end of job. `tf.keras.callbacks.Callback` is an abstract base class and has methods to perform the behavior at different call frequency, such as `on_bath_end`, `on_epoch_end` and so on. So, we adopt the interfaces of `tf.keras.callbacks.Callback` for users to customize the behavior of model in ElasticDL. 
+
+Now we have implemented some modules which are similar to callback in ElasticDL such as `LearingRateScheduler` and `PredictionOutputsProcessor`. And user should write definitions
+in the model definition file for each module like:
+```python
+def custom_model():
+    ...
+
+# Adjust the learning rate according to iteration steps
+def learning_rate_scheduler(model_version):
+    if model_version < 5000:
+        return 0.0003
+    elif model_version < 12000:
+        return 0.0002
+    else:
+        return 0.0001
+
+# Process the prediction outputs for each batch
+class PredictionOutputsProcessor(BasePredictionOutputsProcessor):
+    ...
+```
+There will be multiple definition APIs for users to define different behaviors of model and the interfaces of APIs may be different. It is more convenient for users to define those behaviors using `tf.keras.callbacks.Callback`. 
+
+Some use cases we observed that we want to support are:
+* Case 1: Callback similar to `PredictionOutputsProcessor` that is executed after prediction outputs are made.
+* Case 2: Callback to modulate learning rate as we are currently doing in `LearningRateScheduler`.
+* Case 3: Callback to checkpoint model weights during training.
+* Case 4: Callback to write additional summaries to TensorBoard after each evaluation job completes.
+* Case 5: Callback to perform early stopping when certain conditions are met. For example, the metrics are met after an evaluation job.
+* Case 6: Callback to upload model to remote storage after completing training.
+
+
+To implement callbacks, we need to solve the following problems:
+1. How to define callbacks for the model in ElasticDL.
+2. How to call the methods of callbacks in ElasticDL.
+
+## Define Callbacks for the Model in ElasticDL
+
+In the model definition file, users can add `callbacks` API to define callbacks like:
+```python
+def custom_model():
+    ...
+
+def callbacks():
+    early_stop_callback = EarlyStopCallback()
+    learning_rate_scheduler = LearningRateScheduler()
+    return [early_stop_callback, learning_rate_scheduler]
+
+
+class EarlyStopCallback(tf.keras.callbacks.Callback):
+    def on_batch_end(self, batch, logs=None):
+        ...
+
+class LearningRateScheduler(tf.keras.callbacks.Callback):
```

Comment:
Can we write a complete sample for the class `LearningRateScheduler` here? I'm curious how we get the `model_version` from the parameter in the method `on_batch_start(self, batch, logs=None)`
```
def learning_rate_scheduler(model_version):
    if model_version < 5000:
        return 0.0003
    elif model_version < 12000:
        return 0.0002
    else:
        return 0.0001
```

Reply:
```
import tensorflow as tf
from tensorflow.python.keras import backend as K

def schedule(model_version):
    if model_version < 5000:
        return 0.0003
    elif model_version < 12000:
        return 0.0002
    else:
        return 0.0001

class LearningRateScheduler(tf.keras.callbacks.Callback):
    def __init__(self, schedule, verbose=0):
        super(LearningRateScheduler, self).__init__()
        self.schedule = schedule

    def on_train_batch_begin(self, model_version, logs=None):
        if not hasattr(self.model.optimizer, 'lr'):
            raise ValueError('Optimizer must have a "lr" attribute.')
        lr = self.schedule(model_version)
        K.set_value(self.model.optimizer.lr, lr)
```

I will add the code snippet into the doc.